[CodeGen] Fix cleanup attribute for C89 for-loop init variables

jongmyeong-choi opened the pull request:
In C89, for-init variables have function scope, so cleanup should occur at function exit, not loop exit. This implements deferred cleanup registration for C89 mode while preserving C99+ behavior.

Fixes #154624

Commit:
[CodeGen] Fix cleanup attribute for C89 for-loop init variables

In C89, for-init variables have function scope, so cleanup should occur
at function exit, not loop exit. This implements deferred cleanup
registration for C89 mode while preserving C99+ behavior.

Fixes #154624

Commit:
Apply clang-format to fix code style

efriedma-quic reviewed (commented):
"End of the function" is not really correct. It's still the end of the scope... just the scope in question is different.  This matters if you have, for example, a nested loop.

Probably need to change CodeGenFunction::EmitForStmt to push the scope at the right point.

Commit:
Fix cleanup attribute scope for C89 for-loop variables

Move processDeferredFunctionCleanups() to EmitForStmt to ensure cleanup
happens at correct scope boundary for C89 for-loop init variables.

efriedma-quic reviewed (commented):
Instead of creating a separate cleanup stack and adding special deferred handling for that cleanup stack, can we just skip creating the scope in the first place?  We have `LexicalScope ForScope(*this, S.getSourceRange());`; can we replace that with something like the following?

```
std::optional<LexicalScope> ForScope;
if (getLangOpts().C99 || getLangOpts().CPlusPlus)
  ForScope.emplace(*this, S.getSourceRange());
```

Commit:
Skip creating ForScope for C89 for-loops

Address review feedback: instead of special cleanup handling,
conditionally create LexicalScope only for C99/C++ where
for-init variables have loop scope.

efriedma-quic reviewed (commented):
This is roughly what I expected.  But it looks like there are some build failures on the bot.

Commit:
Fix build errors and clang-format issues

Commit:
clang-format issues

jongmyeong-choi commented:
Thank you for your thorough review. I think this approach is much cleaner than the delayed cleanup 👍 Please let me know if you have any additional comments.

efriedma-quic commented on clang/test/CodeGen/attr-cleanup.c:
```diff
@@ -5,3 +5,26 @@ void g(void) {
 
+// Test for cleanup in for-loop initialization (PR #154624)
+// RUN: %clang_cc1 -std=c89 -emit-llvm %s -o - | FileCheck %s --check-prefix=C89
+// RUN: %clang_cc1 -std=c99 -emit-llvm %s -o - | FileCheck %s --check-prefix=C99
```
Move the RUN lines to the beginning of the file, and delete the redundant RUN line.

Maybe try update_cc_test_checks.py and see if you like the result.

jongmyeong-choi replied:
When you mentioned "delete the redundant RUN line" - were you referring to this line? `// RUN: %clang_cc1 -emit-llvm %s -o %t`

And update_cc_test_checks seems to check the entire IR. I don't have much experience, so I'm wondering if this approach is better. Personally, I feel that localized tests are more intuitive.

efriedma-quic replied:
Yes, that line.

Your choice on update_cc_test_checks; in this case, I could go either way.  Yes, it checks a bunch of stuff that isn't really relevant here, but the fact that it's automated makes it easier to review and update

Commit:
Remove redundant RUN line that only checked compilation success

efriedma-quic reviewed (approved):
LGTM with one minor comment

Commit:
Remove unnecessary whitespace

jongmyeong-choi commented:
> Automation is slightly broken, but:
> 
> > ⚠️ We detected that you are using a GitHub private e-mail address to contribute to the repo.
> > Please turn off [Keep my email addresses private](https://github.com/settings/emails) setting in your account.
> > See [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html#email-addresses) and [LLVM Discourse](https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it) for more information.

Oh, I just turned off the settings. Sorry for bothering you.
